### ISSUE TITLE
RFC: Machine-readable output for wrangler publish

### ISSUE THREAD

**Issue by codewithkristian** [cloudflare/wrangler#1460]
# Situation

When running `wrangler publish`, the user receives the following console output:

```sh
~/src/evergreen/packages/edge$ wrangler publish
✨  Built successfully, built project size is 15 KiB.
🌀  Using namespace for Workers Site "__evergreen-workers_sites_assets"
✨  Success
🌀  Uploading site files
✨  Successfully published your script to https://REDEACTED
💁  Deleting stale files...
```

If something goes wrong, they receive an error string in console output:

```sh
~/src/evergreen/packages/edge$ wrangler publish
✨  Built successfully, built project size is 15 KiB.
Error: ⚠️  Code 10000: Authentication error
```

# Problem

For situations where projects are being deployed _automatically_, via [GitHub Actions](https://github.com/cloudflare/wrangler-action) or something similar, this output is buried inside of the console output. If we were able to _parse_ the information from `wrangler publish`, we would then be able to _pass_ it to other steps in a workflow. 

In https://github.com/cloudflare/wrangler-action/issues/9, I explored potentially parsing the above output via regex, but decided against it because, well, regex. That being said, the advantages of being able to consistently parse `wrangler publish` output is really exciting — we could capture the information and use it to send Slack messages (e.g. "$scriptName has been deployed to $url") or run e2e tests.

In an upcoming project I'm working on with @davidtsong @sgraden on, not having access to this functionality makes it effectively impossible to know _where_ a project has deployed to. Even if a project successfully deploys via GitHub Actions, we can't point them to a URL or even know the name of the script for construct a link to the script in the dashboard (`dash.cloudflare.com/$accountID/workers/$scriptName`).

# Solution

## `wrangler publish --json`

Adding a `--json` flag would return a parseable JSON string that could be passed into `jq` or other tools for getting more information about the published Worker. An example format is shown below:

```sh
$ wrangler publish --json | jq
{
  "success": true,
  "name": "cloudflare-worker-cra",
  "url": "http://cloudflare-worker-cra.signalnerve.workers.dev",
  "environment": "production"
}
```

I would LOVE to have access to `name` and `url` — if the project isn't published under an `environment`, it's possible that field would be confusing or unnecessary (maybe should just be omitted), and `success` is kind of a given if you're getting `name` and `url` back. `success` could be useful given the presence of an `error`:

```sh
$ wrangler publish --json | jq
{
  "success": false,
  "error": {
    "code": 10000,
    "message": "Authentication error"
  }
}
```

# Not Doing

When I floated this idea in our internal chat, there was interest in providing a consistent `--json` flag across all the `wrangler` subcommands. I don't know enough about the other commands to offer a good idea or structure on how to do that, but I'm totally in favor of this work being co-opted for a larger `--json` addition across the repo if we think that's useful. I'm not currently blocked on this (although it would be _really_, _really_ nice to have) so I'm happy to surface this and have the Wrangler team approach it in whatever way works best 👍 

# Open Questions

- [ ] What does the direct response from the Workers API look like? Is it worth it to parse that and show a subset of the response data, or just hand it off directly to the user?
- [ ] What is the correct JSON response for this? Is there a version that can be consistently re-used across multiple commands?

**Comment by ashleymichal** [comment]
i'd like to talk about this in terms of program output (STDOUT) versus diagnostic output (STDERR), where STDOUT acts almost as an api-style contract for the purpose of piping to other processes in a more UNIX-like fashion. with that mindset, what information should be part of the output on a successful publish? what information should be a part of the output in the case of an error?

**Comment by ispivey** [comment]
I wrote some related thoughts about the structure of `tail` output here: https://github.com/cloudflare/wrangler/issues/1368#issuecomment-642600570

**Comment by ashleymichal** [comment]
following up for the sake of today's discussion, i think i like the idea of an  `-o | --output` flag, as suggested by @ispivey in the comment linked above, as well as standardizing the human readable standard output.

the existing progress messages could be moved to STDERR, but the final output (name, url, environment) should be delivered via STDOUT so they can be piped easily. i can see how environment is a useful part of the output; we can call it `default` if no environment is specified in the command.

we can start with wrangler publish and work through the kinks there, since that is absolutely the core of wrangler's functionality, and we can address other commands once we've done that.
